odl/contrib/fom/util.py

```diff
@@ -0,0 +1,437 @@
+        out_img *= out_fh[:, None]
+        out_img *= out_fv[None, :]
+
+        # Inverse trafo
+        # TODO: add `out` parameter to function
+        # TODO: let user give DFT of filters directly
+        conv = image_padded  # Overwrite
+        plan = pyfftw.FFTW(out_img.copy(), conv, axes=(0, 1),
+                           direction='FFTW_BACKWARD', flags=['FFTW_ESTIMATE'],
+                           threads=cpu_count())
+        plan(out_img, conv)
+
+    else:
+        raise ValueError('unsupported `impl` {!r}'.format(impl_in))
+
+    if padding:
+        return conv[padding:-padding, padding:-padding]
+    else:
+        return conv
+
+
+def similarity_score(x, y, c):
+    """Pointwise similarity measure of two arrays.
+
+    The score is computed as ::
+
+        S(x, y, c) = (2 * x * y + c ** 2) / (x ** 2 + y ** 2 + c ** 2)
+
+    This is formula (3) in `[Rei+2016] <https://arxiv.org/abs/1607.06140>`_,
+    except for the usage of `c ** 2` instead of `c`.
+
+    Parameters
+    ----------
+    x, y : array-like
+        The arrays to score on pointwise similarity. They are assumed to
+        have nonnegative entries only.
+    c : positive float
+        Constant determining the score of maximally dissimilar values.
+        Smaller constant means higher penalty for dissimilarity.
+        See Notes for details.
+
+    Returns
+    -------
+    score : `numpy.ndarray`
+        The pointwise score of similarity.
+
+    Notes
+    -----
+    For :math:`x = y`, the returned score is 1 independent of :math:`c`.
+    If :math:`x = 0`, the score is equal to
+
+    .. math::
+        \\frac{c^2}{y^2 + c^2} = \\frac{1}{(y/c)^2 + 1}.
+
+    Hence, :math:`c` is the :math:`y`-value at which the score
+    drops to :math:`1 / 2` for :math:`x = 0`. In other words, the smaller
+    :math:`c` is chosen, the more dissimilarity is penalized.
+
+    The value should be chosen in the same order of magnitude as "typical"
+    values of :math:`y`.
+
+    References
+    ----------
+    [Rei+2016] Reisenhofer, R, Bosse, S, Kutyniok, G, and Wiegand, T.
+    *A Haar Wavelet-Based Perceptual Similarity Index for Image Quality
+    Assessment*. arXiv:1607.06140 [cs], Jul. 2016.
+    """
+    x = np.asarray(x)
+    y = np.asarray(y)
+    c = float(c)
+    return (2 * x * y + c ** 2) / (x ** 2 + y ** 2 + c ** 2)
+
+
+def logistic(x, a):
+    """The logistic function.
+
+    It is defined as ::
+
+        l_a(x) = 1 / (1 + exp(-a * x))
+    """
+    return 1 / (1 + np.exp(-a * x))
+
+
+def inv_logistic(x, a):
+    """The inverse of the logistic function.
+
+    It is defined as ::
+
+        l_a^{-1}(x) = (log(x) - log(1 - x)) / a
+    """
+    return (np.log(x) - np.log(1 - x)) / a
+
+
+def similarity_map(img1, img2, axis, c, a):
+    """Local similarity score for directional features along an axis.
+
+    Parameters
+    ----------
+    img1, img2 : array-like
+        The images to compare. They must have equal shape.
+    axis : {0, 1}
```

Comment:
Any reason this is 2d only?

Reply:
Well I didn't bother to do nD for now. Wouldn't be a big problem. For now I added TODO comments.